packages/mui-joy/src/utils/useSlot.ts

```diff
@@ -0,0 +1,165 @@
+import * as React from 'react';
+import { ClassValue } from 'clsx';
+import { unstable_useForkRef as useForkRef } from '@mui/utils';
+import {
+  appendOwnerState,
+  resolveComponentProps,
+  unstable_mergeSlotProps as mergeSlotProps,
+} from '@mui/base/utils';
+
+export type WithCommonProps<T> = T & {
+  className?: string;
+  style?: React.CSSProperties;
+  ref?: React.Ref<any>;
+};
+
+type EventHandlers = Record<string, React.EventHandler<any>>;
+
+type ExtractComponentProps<P> = P extends infer T | ((ownerState: any) => infer T) ? T : never;
+
+/**
+ * An internal function to create a Joy UI slot.
+ *
+ * This is an advanced version of MUI Base `useSlotProps` because Joy UI allows leaf component to be customized via `component` prop
+ * while MUI Base does not need to support leaf component customization.
+ *
+ * @param {string} name: name of the slot
+ * @param {object} parameters
+ * @returns {[Slot, slotProps]} The slot's React component and the slot's props
+ *
+ * Note: the returned slot's props
+ * - will never contain `component` prop.
+ * - might contain `as` prop.
+ */
+export default function useSlot<
+  T extends string,
+  ElementType extends React.ElementType,
+  SlotProps,
+  OwnerState extends {},
+  ExternalSlotProps extends { component?: React.ElementType },
+  ExternalForwardedProps extends {
+    component?: React.ElementType;
+    components?: { [k in T]?: React.ElementType };
+    componentsProps?: {
+      [k in T]?:
+        | WithCommonProps<ExternalSlotProps>
+        | ((ownerState: OwnerState) => WithCommonProps<ExternalSlotProps>);
+    };
+  },
+  AdditionalProps,
+  SlotOwnerState extends {},
+>(
+  /**
+   * The slot's name. All Joy UI components should have `root` slot.
+   *
+   * If the name is `root`, the logic behaves differently from other slots,
+   * e.g. the `externalForwardedProps` are spread to `root` slot but not other slots.
+   */
+  name: T,
+  parameters: (T extends 'root' ? { ref: React.ForwardedRef<any> } : {}) & {
+    /**
+     * The slot's className
+     */
+    className: ClassValue | ClassValue[];
+    /**
+     * The slot's default styled-component
+     */
+    elementType: ElementType;
+    /**
+     * The component's ownerState
+     */
+    ownerState: OwnerState;
+    /**
+     * The `other` props from the consumer. It has to contain `component`, `components`, and `componentsProps`.
```

Comment:
All of these three must always be there? For example, `CircularProgress` is missing `components` prop. We should add `components` prop to `CircularProgress`?

Reply:
> We should add components prop to CircularProgress?

Do you mean the types or the logic?

Reply:
If we decide to add the type, shouldn't we add the logic too? We don't handle `components` prop in Circular Progress, but should we? That's my question.

Reply:
> If we decide to add the type, shouldn't we add the logic too? We don't handle components prop in Circular Progress, but should we? That's my question.

Got your point. Yes, all Joy UI components that have more than root slot should receive `components` and `componentsProps`.

Reply:
Gottcha, I will work on that.

Reply:
We already renamed these props to `slots` and `slotsProps`, and we are adding these new props in Material UI. Is there apply to do the same for Joy UI soon?

Reply:
@mnajdova Benny is working on that in a separate PR (WIP) to move from `components` to `slots` prop.